### ISSUE TITLE
Investigate bidirectional / unidirectional connections

### ISSUE THREAD

**Issue by cwgoes** [cosmos/ics#72]
- Can we construct a bidirectional connection from two unidirectional ones?
- How would the handshake work in that case?
- Will this simplify the protocol design?

**Comment by mossid** [comment]
Current state machine for each chain:

```
           +-----------+
           |           |
    +------>  Neutral  +<-----+
    |      |           |      |
Timeout    +-----+-----+    Timeout
    |            |            |
    |            |            |
    |   +--------+--------+   |
    |   |                 |   |
    |   |                 |   |
    |   v                 v   |
  +-+---+-----+     +-----+---+-+
  |           |     |           |
  | OpenInit  |     |  OpenTry  |
  |           |     |           |
  +-----+-----+     +-----+-----+
        |                 |
        |                 |
        v                 v
  +-----+-----+     +-----+-----+
  |           |     |           |
  |  OpenAck  |     |  OpenConf |
  |           |     |           |
  +-----+-----+     +-------+---+
        |                   |
        |  +------------+   |
        |  |            |   |
        +-->   Open     <---+
           |            |
           +------------+

```

A unidirectional connection is a chain can only send packets throught the connection but cannot receive, and vice versa on the other chain. It is actually defined as a pair of `(writeonly partial connection on chain A, readonly partial connection on chain B)`.

I don't think we can merge existing two unidirectional connections to build a bidirectional connection. Those two unidirectional connections will be defined as 
```
(Readonly on chain A with ID1, Writeonly on chain B with ID2)
(Writeonly on chain A with ID3, Readonly on chain B with ID4)
```
and then we have to generate `ID1+3` on chain A, `ID2+4` on chain B and allocate it to the new bidirectional connection, and I think it can invalidate the existing assets on the other chain, cause massive change on the stored identifiers, etc.

Instead we can extend our existing negotiation protocol to handle the unidirectional handshake, so the intermediate states(Neutral, OpenInit, OpenTry) can be treated as `read-only` or `write-only`. For example at `OpenInit` state the connection can not ensure that the message it is reading is intended to be sent to itself. In this state we can conceptualize that it is composed of two unidirectional(one readonly, one writeonly) connections, which might not directing same connection(yet), but sharing a same connection ID. 

So we can make the connection still read or write on the intermediate states, but the application logic should ensure that the packet it is trying to sending is not requiring bidirectional connection. For example, a simple message broadcasting application does not have to ensure that the handshake is complete, a simple state reading application also does not have to, but a coin sending application should ensure that the handshake is complete. 

In summary: instead of merging two unidirectional connections, we enable to send or receive packets over a connection whose handshake it not complete, without ensuring the reading and writing part of the connection is compatible with each other, which essentially works as two unidirectional connection(one readonly one writeonly) sharing same connection ID.

**Comment by cwgoes** [comment]
> and then we have to generate ID1+3 on chain A, ID2+4 on chain B and allocate it to the new bidirectional connection, and I think it can invalidate the existing assets on the other chain, cause massive change on the stored identifiers, requires migration of existing packets, etc.

I don't quite understand this. Identifiers can be prefixed with an indicator for unidirectional or bidirectional connection type, so conflicts aren't a problem, and the modules which owned the unidirectional connections can now own the bidirectional connection - it doesn't seem necessary to tie *assets* themselves to a particular connection identifier (although we *do* need to flesh out the permissioning scheme around module connection usage more concretely).

> Instead we can extend our existing negotiation protocol to handle the unidirectional handshake, so the intermediate states(Neutral, OpenInit, OpenTry) can be treated as `read-only` or `write-only`.

We *could* do this, but I'm not sure this actually provides what we want out of unidirectional connections - for example, unidirectional connections could be used in a one-to-many broadcast scenario, where the receiving connections should be creatable without any confirmation from the broadcasting chain. 

> For example, a simple message broadcasting application does not have to ensure that the handshake is complete, a simple state reading application also does not have to, but a coin sending application should ensure that the handshake is complete.

Yes, modules will need to assert the correct channel / connection type.

**Comment by cwgoes** [comment]
@mossid & I discussed in a whiteboarding session. We are in agreement that the light client verification semantics can be separated from ordering and queue-identifier-association semantics so that two unidirectional verification sessions in opposite directions can be used for bidirectional verification (although we don't think it makes sense to call this merging two unidirectional connections, for reasons of avoiding handshake duplication later on).

**Comment by cwgoes** [comment]
![one](https://i.imgur.com/Cp7AP6e.jpg)

![two](https://i.imgur.com/uI7VB4w.jpg)

![three](https://i.imgur.com/r10zEcK.jpg)

**Comment by cwgoes** [comment]
We will write up an alternative "connection semantics" ICS later this week.

**Comment by mossid** [comment]
![a68ebe38-b2dc-4552-979e-e5e281f380cf](https://user-images.githubusercontent.com/18329646/56363307-2efcb400-61ec-11e9-91ad-a4cce7d89679.jpeg)

**Comment by cwgoes** [comment]
I think we've reached a solid consensus on the abstraction layers here, at least for the time being.
